### PR TITLE
Savestates: Implement reload savestate shortcut and some bugfixes

### PR DESCRIPTION
**Pull request by elad335**
Ctrl+R to reload savestate during gameplay if it exists and is compatible with the current RPCS3 version.
You need the "Suspend Emulation Savestate Mode" setting on in the config file to use it.
The reason this setting is not on by default is that there is a risk HDD0 state has been changed by the game when playing after the savestate has been created in a way that is not compatible with the created savestate file. (because HDD0 state is not saved by default at the moment, so use with caution)

In addition save RSX semaphore waiting state, for synchronization should not be harmed when reloading.

### CODE REVIEW

File: rpcs3/rpcs3qt/gs_frame.cpp [hunk 1]
```diff
@@ -247,9 +247,22 @@ void gs_frame::keyPressEvent(QKeyEvent *keyEvent)
...
+			{
+				Emu.Run(true);
+				return;
+			}
+			case system_state::paused:
+			{
+				Emu.Resume();
+				return;
+			}
+			default: Emu.Pause(); break;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/gs_frame.cpp:264]: Might as well put this in a scope as well

File: rpcs3/rpcs3qt/gs_frame.cpp [hunk 2]
```diff
@@ -261,21 +274,42 @@ void gs_frame::keyPressEvent(QKeyEvent *keyEvent)
...
 	case Qt::Key_R:
-		if (keyEvent->modifiers() == Qt::ControlModifier && !m_disable_kb_hotkeys)
+		if (keyEvent->modifiers() == Qt::ControlModifier && !m_disable_kb_hotkeys && !g_cfg.savestate.suspend_emu && !Emu.GetTitleID().empty() && (Emu.IsRunning() || Emu.GetStatus() == system_state::paused))
 		{
-			switch (Emu.GetStatus())
-			{
-			case system_state::ready:
+			extern bool is_savestate_compatible(const fs::file& file);
+
+			const std::string save_dir = fs::get_cache_dir() + "/savestates/";\
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/gs_frame.cpp:281]:
```suggestion
			const std::string save_dir = fs::get_cache_dir() + "/savestates/";
```

File: rpcs3/rpcs3qt/gs_frame.cpp [hunk 3]
```diff
@@ -261,21 +274,42 @@ void gs_frame::keyPressEvent(QKeyEvent *keyEvent)
...
+
+			std::string savestate_path;
+			u64 mtime = 0;
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/gs_frame.cpp:284]: What is this for?

File: rpcs3/rpcs3qt/gs_frame.cpp [hunk 4]
```diff
@@ -261,21 +274,42 @@ void gs_frame::keyPressEvent(QKeyEvent *keyEvent)
...
+
+			for (auto&& entry : fs::dir(save_dir))
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/gs_frame.cpp:286]: Why not do this in the new utils file?

File: rpcs3/Emu/System.cpp [hunk 5]
```diff
@@ -129,6 +130,36 @@ SERIALIZATION_VER(cellAudioOut, 22,                             1)
 
 #undef SERIALIZATION_VER
 
+bool is_savestate_version_compatible(const std::vector<std::pair<u16, u16>>& data, bool log)
```
Comment (Megamouse) [inline rpcs3/Emu/System.cpp:133]: Cant this be part of the new utils file?

File: rpcs3/rpcs3qt/gs_frame.cpp (hunk 4 above)
Comment (elad335) [reply rpcs3/rpcs3qt/gs_frame.cpp:289]: It is a very specific situation and is not directly related to savestates. (picking up the most recent file that is)

Comment (Megamouse) [reply rpcs3/rpcs3qt/gs_frame.cpp:289]: A random directory scanner shouldn't be in a game window class

Comment (Megamouse) [reply rpcs3/rpcs3qt/gs_frame.cpp:289]: Everything after m_disable_kb_hotkeys should be in a savestate function

File: rpcs3/Emu/savestate_utils.cpp [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+	for (serial_ver_t& ver : s_serial_versions)
+	{
+		if (std::exchange(ver.used, false))
+		{
+			used_serial.emplace_back(&ver - s_serial_versions.data(), *ver.compatible_versions.rbegin());
+		}
+	}
+
+	return used_serial;
+}
```
Comment (Megamouse) [inline rpcs3/Emu/savestate_utils.cpp:156]: Missing newline

**Pull request merged**